File: presentation/src/main/java/will/shiro/desafiopicpay/util/base/BaseActivity.java
package will.shiro.desafiopicpay.util.base;

import android.app.Fragment;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.CallSuper;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;

public abstract class BaseActivity extends AppCompatActivity implements HasAndroidInjector {

    @Inject
    DispatchingAndroidInjector<Fragment> fragmentDispatchingAndroidInjector;

    @Inject
    DispatchingAndroidInjector<Object> androidInjector;

    @Override
    public AndroidInjector<Object> androidInjector() {
        return androidInjector;
    }

    protected abstract BaseViewModel baseViewModel();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        subscribeUi();
    }

    @CallSuper
    protected void subscribeUi() {
        baseViewModel().getToast().observe(this, new Observer<Event<String>>() {
            @Override
            public void onChanged(Event<String> stringEvent) {
                String string = stringEvent.getContentIfNotHandled();
                if (string != null) {
                    Toast.makeText(getBaseContext(), string, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }
}

File: presentation/src/main/java/will/shiro/desafiopicpay/util/base/BaseActivity.kt
package will.shiro.desafiopicpay.util.base

import android.os.Bundle
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import dagger.android.AndroidInjection
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import will.shiro.desafiopicpay.util.extensions.observeEvent
import javax.inject.Inject

abstract class BaseActivity2 : AppCompatActivity(), HasAndroidInjector {

    @Inject
    protected lateinit var fragmentInjector: DispatchingAndroidInjector<Fragment>

    @Inject
    lateinit var androidInjector: DispatchingAndroidInjector<Any>

    override fun androidInjector(): AndroidInjector<Any> = androidInjector

    abstract val baseViewModel: BaseViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        subscribeUi()
    }

    @CallSuper
    open fun subscribeUi() {
        baseViewModel.toast.observeEvent(this, ::onNextToast)
    }

    private fun onNextToast(text: String?) {
        text?.let {
            Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
        }
    }
}
